# Characterize errors in consensus calling

| Tree | Rep | Hap | n_mut | Path | Notes | Potential fix |
| ---- | --- | --- | ----- | ---- | ----- | ------------- |
|RSV|4|10|1|[comparison_output](heuristic7_output/rsv4000_rep4_10hap_1snp_consensus_heuristic_merged/consensus_to_true_comparison_heuristic_kl_gap_merged_heuristic7.tsv)|Two entries in `log_likelihood_ratio_by_num_haplotypes_assigned` and both only have one candidate allele (thus both have 0 likelihood ratio)|Might need to double if it's really true but in cases like this where each read group only has one allele are has different alleles for different groups, there must be variations near the current positions that differentiate these read groups. I can potentially use that information. In this case, IGV shows that most reads supporting allele A also support an allele C 38 bases upstream, while reads supporting G also support A at 38 bases upstream. We know that the target allele has an A at 38 bases upstream so the allele at the current position should be G|
|RSV|7|10|1|[comparison_output](heuristic7_output/rsv4000_rep7_10hap_1snp_consensus_heuristic_merged/consensus_to_true_comparison_heuristic_kl_gap_merged_heuristic7.tsv)|Target haplotype (0.05 abundance) and abother haplotype (0.05 abandance) both have the same set of two candidate alleles.|Not sure. Alleles could go either way|
|~~RSV~~|~~9~~|~~10~~|~~1~~|[comparison_output](heuristic7_output/rsv4000_rep9_10hap_1snp_consensus_heuristic_merged/consensus_to_true_comparison_heuristic_kl_gap_merged_heuristic7.tsv)|~~Both likelihood ratios support the same allele. The allele was assigned but are both at low ratio~~|~~Ensure tha if there is a single allele supported by tge likelihood ratios make sure the accumulated ratio is greater than threshold to assign allele.~~ *Fixed*|
|RSV|5|10|5|[comparison_output](heuristic7_output/rsv4000_rep5_10hap_5snp_consensus_heuristic_merged/consensus_to_true_comparison_heuristic_kl_gap_merged_heuristic7.tsv)|Likely to be a sequencing error|Increase the minimum depth to count allele or better way to detect sequencing error?|
|~~RSV~~|~~7~~|~~10~~|~~5~~|[comparison_output](heuristic7_output/rsv4000_rep7_10hap_5snp_consensus_heuristic_merged/consensus_to_true_comparison_heuristic_kl_gap_merged_heuristic7.tsv)|~~Same as `rsv_9_10_1`~~|~~Same as `rsv_9_10_1`~~ *Fixed*|
|RSV|8|10|10|[comparison_output](heuristic7_output/rsv4000_rep8_10hap_10snp_consensus_heuristic_merged/consensus_to_true_comparison_heuristic_kl_gap_merged_heuristic7.tsv)|Same as `rsv_5_10_5`|Same as `rsv_5_10_5`|
|RSV|9|10|10|[comparison_output](heuristic7_output/rsv4000_rep9_10hap_10snp_consensus_heuristic_merged/consensus_to_true_comparison_heuristic_kl_gap_merged_heuristic7.tsv)|(pos 11536, KJ672469.1) Not enough confidence to call true variant allele|Lower threshold?|
|RSV|9|10|10|[comparison_output](heuristic7_output/rsv4000_rep9_10hap_10snp_consensus_heuristic_merged/consensus_to_true_comparison_heuristic_kl_gap_merged_heuristic7.tsv)|(pos 8710, KU950541.1) Same as `rsv_9_10_10_11536_KJ672469.1`|Same as `rsv_9_10_10_11536_KJ672469.1`|
|RSV|9|10|10|[comparison_output](heuristic7_output/rsv4000_rep9_10hap_10snp_consensus_heuristic_merged/consensus_to_true_comparison_heuristic_kl_gap_merged_heuristic7.tsv)|(pos 11734, KU950541.1)Same as `rsv_5_10_5`|Same as `rsv_5_10_5`|
|RSV|9|10|10|[comparison_output](heuristic7_output/rsv4000_rep9_10hap_10snp_consensus_heuristic_merged/consensus_to_true_comparison_heuristic_kl_gap_merged_heuristic7.tsv)|(pos 9316, MN365585.1) Same as `rsv_7_10_1`|Same as `rsv_7_10_1`|
|HIV|10|10|10|[comparison_output](heuristic7_output/hiv20000_rep10_10hap_10snp_consensus_heuristic_merged/consensus_to_true_comparison_heuristic_kl_gap_merged_heuristic7.tsv)|(pos 8502,8504,8526,8533,8576,8594,8596,8600,8601,8622,8638,8780,8799,8805,8833,8834,8896,8977, B_US_2017_2026_EMDRn_1_MW753727) All have allele-determining read groups but the group support multiple alleles that are not sequencing errors. Turned out the chaining scores of reads that span the region are all very low.|Deal with allele-determining read groups more cleverly. Perhaps use read to reference similarity to determine allele|
|HIV|10|10|5|[comparison_output](heuristic7_output/hiv20000_rep10_10hap_5snp_consensus_heuristic_merged/consensus_to_true_comparison_heuristic_kl_gap_merged_heuristic7.tsv)|(pos 8469,8474,8477,8488,8502,8504,8526,8533,8576,8594,8596,8600,8601,8622,8638,8896,8967,8968, B_US_2017_2026_EMDRn_1_MW753727) Same as `hiv_10_10_10`| Same as `hiv_10_10_10`|
|SARS|5|2|5|[comparison_output](heuristic7_output/sars20000_rep5_2hap_5snp_consensus_heuristic_merged/consensus_to_true_comparison_heuristic_kl_gap_merged_heuristic7.tsv)|(pos 6514,21754, Scotland_QEUH_345F3FD_2022_OV839107_1_2022_01_26) Insertions induced| Extend consensus calling to deal with indels |
|SARS|5|3|5|[comparison_output](heuristic7_output/sars20000_rep5_2hap_5snp_consensus_heuristic_merged/consensus_to_true_comparison_heuristic_kl_gap_merged_heuristic7.tsv)| (all three haplpotype at the same aligned position) | Extend consensus calling to deal with indels |

